### PR TITLE
feat: JS tracer callbacks

### PR DESCRIPTION
**Pull request by Wodann**

### CODE REVIEW

File: crates/edr_evm_napi/src/tracer/js_tracer.rs [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+}
+
+impl JsTracer {
+    /// Constructs an instance from the provided callbacks.
+    pub fn new(env: &Env, callbacks: TracingCallbacks) -> napi::Result<Self> {
+        let before_call_fn = ThreadsafeFunction::create(
+            env.raw(),
+            unsafe { callbacks.before_call.raw() },
+            0,
+            |mut ctx: ThreadSafeCallContext<BeforeCallEvent>| {
```
Comment (agostbiro) [inline crates/edr_evm_napi/src/tracer/js_tracer.rs:53]: I find it difficult to parse what's going on in these function bodies. 😅 Could you add some high level comments both in the before and after calls explaining the flow please?

**Review by agostbiro — CHANGES_REQUESTED** [review]
Thanks for this difficult PR @Wodann ! I added some comments, but looks good overall

**Comment by agostbiro** [comment]
There are two CI failures:

- The JS [Lint](https://github.com/NomicFoundation/hardhat/actions/runs/6552758788/job/17796753577?pr=4487#logs)
- Some state root errors, but weirdly only in [Test provider (ubuntu-latest, Node 18.15, dual)](https://github.com/NomicFoundation/hardhat/actions/runs/6552758793/job/17796817860?pr=4487#logs). Assuming these are fixed in `rethnet/main`?

File: crates/edr_evm_napi/src/tracer/js_tracer.rs [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+        Ok(Self {
+            before_call_fn,
+            after_call_fn,
+            pending_before: None,
+        })
+    }
+
+    fn validate_before_message(&mut self) {
+        if let Some(message) = self.pending_before.take() {
+            let (sender, receiver) = channel();
```
[the remark this answers is not in the document]
Comment (Wodann) [reply crates/edr_evm_napi/src/tracer/js_tracer.rs:349]: Communicating between two threads: the thread-safe function and EDR

File: crates/edr_evm_napi/src/tracer.rs [hunk 3]
```diff
@@ -0,0 +1,28 @@
+mod js_tracer;
+
+use edr_evm::{blockchain::BlockchainError, state::StateError, SyncInspector};
+use napi::Env;
+use napi_derive::napi;
+
+use self::js_tracer::{JsTracer, TracingCallbacks};
+
+#[napi]
+pub struct Tracer {
```
[the remark this answers is not in the document]
Comment (Wodann) [reply crates/edr_evm_napi/src/tracer.rs:10]: To generate napi bindings. The `JsTracer` is a Rust-only struct

File: crates/edr_evm_napi/src/runtime.rs [hunk 4]
```diff
@@ -26,14 +28,13 @@ pub async fn dry_run(
     transaction: TransactionRequest,
     block: BlockConfig,
     with_trace: bool,
+    tracer: Option<&Tracer>,
```
[the remark this answers is not in the document]
Comment (Wodann) [reply crates/edr_evm_napi/src/runtime.rs:31]: Given that most of the `napi` functions are temporary, I haven't bothered reducing the number of arguments.

If this code was anywhere else, I'd follow your recommendation

File: crates/edr_evm_napi/src/trace.rs [hunk 5]
```diff
@@ -86,6 +97,38 @@ impl TracingMessage {
...
+    fn try_cast(self) -> napi::Result<BeforeMessage> {
+        let to = self.to.map(|to| Address::from_slice(to.as_ref()));
+        let data = Bytes::copy_from_slice(self.data.into_value()?.as_ref());
+        let value = BigInt::try_cast(self.value)?;
+        let code_address = self
+            .code_address
+            .map(|code_address| Address::from_slice(code_address.as_ref()));
+        let code = self
+            .code
+            .map(|code| {
```
[the remark this answers is not in the document]
Comment (Wodann) [reply crates/edr_evm_napi/src/trace.rs]: Addressed in https://github.com/NomicFoundation/hardhat/pull/4487/commits/3a2060caf66245414be70db0444625ad8379c2ee

File: crates/edr_evm_napi/src/tracer/js_tracer.rs [hunk 6]
```diff
@@ -0,0 +1,490 @@
...
+    after_call_fn: ThreadsafeFunction<AfterCallEvent>,
+    pending_before: Option<BeforeMessage>,
+}
+
+impl JsTracer {
+    /// Constructs an instance from the provided callbacks.
+    pub fn new(env: &Env, callbacks: TracingCallbacks) -> napi::Result<Self> {
+        let before_call_fn = ThreadsafeFunction::create(
+            env.raw(),
+            unsafe { callbacks.before_call.raw() },
```
[the remark this answers is not in the document]
Comment (Wodann) [reply crates/edr_evm_napi/src/tracer/js_tracer.rs:52]: Addressed in https://github.com/NomicFoundation/hardhat/pull/4487/commits/4f7dd4ae886923393968a8b060ab71b2a9607b1c

File: crates/edr_evm_napi/src/tracer/js_tracer.rs (hunk 1 above)
Comment (Wodann) [reply crates/edr_evm_napi/src/tracer/js_tracer.rs:54]: Hopefully, addressed in https://github.com/NomicFoundation/hardhat/pull/4487/commits/4f7dd4ae886923393968a8b060ab71b2a9607b1c

**Review by agostbiro — APPROVED** [review]
Thanks, LGTM!

**Pull request merged**